daml-assistant/daml-project-config/DAML/Project/Consts.hs

```diff
@@ -157,9 +168,20 @@ withProjectRoot mbProjectDir (ProjectCheck cmdName check) act = do
             when check $ do
                 hPutStrLn stderr (cmdName <> ": Not in project.")
                 exitFailure
-            act pure
+            act mbProjectPath pure
         Just projectPath -> do
             projectPath <- canonicalizePath projectPath
-            withCurrentDirectory projectPath $ act $ \f -> do
+            withCurrentDirectory projectPath $ act mbProjectPath $ \f -> do
                 absF <- canonicalizePath (previousCwd </> f)
                 pure (projectPath `makeRelative` absF)
+
+-- | Same as 'withProjectRoot' but always requires project root.
+withExpectProjectRoot
```

Comment:
I’m not sure I understand this function. Doesn’t setting `check` to `True` accomplish this? I don’t see how we can fall into the `Just projectPath` of `withProjectRoot` but have `mbProjectPath` be `Nothing`.

Reply:
`withExpectProjectRoot` is internally the same as `withProjectRoot` with `check` set to `True`.If `check` is `True` then `withProjectRoot` will always provide `Just pPath` (or terminate the program). `withExpectProjectRoot` just makes this explicit in the types, so that the use-site doesn't have to worry about this.

Reply:
I’m still confused. It looks like the only case where you use the additional argument in `withProjectRoot` is `withExpectProjectRoot` and there it will always be `Just`.

Reply:
I'm afraid I don't understand the question.

The motivation is this:

Before the use-site looked like this
```
    withProjectRoot' projectOpts $ \_ -> do
        projectPath <- maybe getProjectPath (pure . Just . unwrapProjectPath) (projectRoot projectOpts)
        case projectPath of
            ...
```
I.e. the project path discovery had to be repeated and there was always a case match on `projectPath`, which could be `Nothing` or `Just ...`.

Now it looks like this
```
    withExpectProjectRoot (projectRoot projectOpts) "daml test" $ \pPath _ -> do
```
I.e. the project path is directly available without repeating the discovery or having to do additional case matches.

We can do this because

> It's an error to run damlc test outside a project if --files is not specified.

Reply:
Ah I see, I missed the change in `runTestsInProjectOrFiles`. LGTM :+1: